odl/space/npy_tensors.py

```diff
@@ -0,0 +1,1992 @@
+             [-9.0, 5.0, -9.0]]
+        )
+
+        """
+        arr = self.data[indices]
+        if np.isscalar(arr):
+            return arr
+        else:
+            space_constructor = type(self.space)
+            if (self.order in ('C', 'F') and
+                    arr.flags[self.order + '_CONTIGUOUS']):
+                out_spc_order = self.order
+            else:
+                # To preserve the array view for non-contiguous slices,
+                # we need to use 'A' for the space in that case.
+                out_spc_order = 'A'
+            space = space_constructor(
+                arr.shape, dtype=self.dtype, order=out_spc_order,
+                exponent=self.space.exponent, weighting=self.space.weighting)
+            return space.element(arr)
+
+    def __setitem__(self, indices, values):
+        """Implement ``self[indices] = values``.
+
+        Parameters
+        ----------
+        indices : index expression
+            Integer, slice or sequence of these, defining the positions
+            of the data array which should be written to.
+        values : scalar, array-like or `NumpyTensor`
+            The value(s) that are to be assigned.
+
+            If ``index`` is an integer, ``value`` must be a scalar.
+
+            If ``index`` is a slice or a sequence of slices, ``value``
+            must be broadcastable to the shape of the slice.
+
+        Examples
+        --------
+        For 1d spaces, entries can be set with scalars or sequences of
+        correct shape:
+
+        >>> space = odl.rn(3)
+        >>> x = space.element([1, 2, 3])
+        >>> x[0] = -1
+        >>> x[1:] = (0, 1)
+        >>> x
+        rn(3).element([-1.0, 0.0, 1.0])
+
+        In higher dimensions, the Numpy assignment and broadcasting rules
+        apply. It is also possible to use tensors in other spaces:
+
+        >>> space = odl.rn((2, 3))
+        >>> x = space.element([[1, 2, 3],
+        ...                    [4, 5, 6]])
+        >>> x[0, 1] = -1
+        >>> x
+        rn((2, 3)).element(
+            [[1.0, -1.0, 3.0],
+             [4.0, 5.0, 6.0]]
+        )
+        >>> space = odl.tensor_space((2, 2), dtype='short')
+        >>> y = space.element([[-1, 2],
+        ...                    [0, 0]])
+        >>> x[:, :2] = y
+        >>> x
+        rn((2, 3)).element(
+            [[-1.0, 2.0, 3.0],
+             [0.0, 0.0, 6.0]]
+        )
+        >>> x[0, 1:] = [7, 8]
+        >>> x
+        rn((2, 3)).element(
+            [[-1.0, 7.0, 8.0],
+             [0.0, 0.0, 6.0]]
+        )
+        >>> x[:] = np.array([[0, 0, 0],
+        ...                  [1, 1, 1]])
+        >>> x
+        rn((2, 3)).element(
+            [[0.0, 0.0, 0.0],
+             [1.0, 1.0, 1.0]]
+        )
+        >>> x[:, ::2] = -2.
+        >>> x
+        rn((2, 3)).element(
+            [[-2.0, 0.0, -2.0],
+             [-2.0, 1.0, -2.0]]
+        )
+
+        Be aware of unsafe casts and over-/underflows, there
```

Comment:
Yeah, we had this before and I kept it, but I also don't think it's super important.